feat: transactions page

lcsfelix opened the pull request.

lcsfelix committed:
feat: transactions page

lcsfelix committed:
fix: remove unused txHash function

lcsfelix committed:
refactor: rewrite transactions sql queries

lcsfelix committed:
fix: use left instead of inner join on txs

RafaelTaranto commented on new-lamassu-admin/src/components/table/TableRow.js:
```diff
@@ -24,7 +24,7 @@ const useStyles = makeStyles({
   },
   lg: {
-    extend: info2,
+    // extend: info2,
```
Any reason why this was commented out?

lcsfelix replied:
Hum... I'm sure there was, but now uncommenting doesn't seem to make a difference, so I'll revert it.

joshmh replied:
Why not remove it, in that case?

RafaelTaranto commented on new-lamassu-admin/src/components/LogsDownloaderPopper.js:
```diff
@@ -132,36 +131,26 @@ const LogsDownloaderPopover = ({ id, open, anchorEl, onClose, logsResponse, ...p
-    setRadioButtons(toInteger(event.target.value))
+    const radio = toInteger(event.target.value)
+    setRadioButtons(radio)
+    if (radio === 0) setRange({ from: null, to: null })
```
why not just use boolean?

lcsfelix replied:
The RadioGroup component is prepared for more than 2 options. Using the `value` identifier even when only two options are used is more consistent.

RafaelTaranto replied:
better to have name values, then.

lcsfelix replied:
Fixed in https://github.com/lamassu/lamassu-server/pull/336/commits/58c08e39d6c46210864dc97c5ae25ec7fae489b7

RafaelTaranto commented on new-lamassu-admin/src/components/expandable-table/ExpTable.js:
```diff
@@ -28,7 +28,7 @@ const ExpRow = ({ id, columns, details, sizes, expanded, className, expandRow, .
   const detailsRowClasses = {
     [classes.detailsRow]: true,
-    [classes.hideDetailsRow]: expanded
+    [classes.hideDetailsRow]: !expanded
```
Prevent the component from being mounted instead of using a class 

`{expanded && (<Tr....)`

RafaelTaranto commented on new-lamassu-admin/src/components/expandable-table/ExpTable.js:
```diff
@@ -28,7 +28,7 @@ const ExpRow = ({ id, columns, details, sizes, expanded, className, expandRow, .
+    [classes.hideDetailsRow]: !expanded
   }
 
   return (
```
Why are you using uuidv1 on all the components here? You assign key for the components only for the react internals, rule of thumb is using `idx`. If you generate the key every time it will be as if you're handing a new component every time.

RafaelTaranto replied:
Also I don't even see uuid lib on package.json, so it wouldn't work on production.

RafaelTaranto commented on new-lamassu-admin/src/components/expandable-table/ExpTable.js:
```diff
@@ -41,7 +41,8 @@ const ExpRow = ({ id, columns, details, sizes, expanded, className, expandRow, .
+        </Td>
       </Tr>
       <Tr className={classnames(detailsRowClasses)}>
         <Td size={mainWidth}>
```
rowStates logic is awfully complex here, if we need to keep only one row open why not just have a id state with the open component? Then we don't have to keep creating a new list and mapping everything on that list to false.

RafaelTaranto commented on new-lamassu-admin/src/components/expandable-table/ExpTable.js:
```diff
@@ -28,7 +28,7 @@ const ExpRow = ({ id, columns, details, sizes, expanded, className, expandRow, .
 
```
This component has some serious performance issues, I'll try to lay it all out here. Some comments are not on the right place because I can only review the current PR

RafaelTaranto commented on new-lamassu-admin/src/pages/ServerLogs.js:
```diff
@@ -108,11 +109,34 @@ const Logs = () => {
+        type: 'text/plain;charset=utf-8'
+      })
+      FileSaver.saveAs(blob, `${formatDateFile(range.from)}_${formatDateFile(range.to)}_server`)
+    }
```
I've seen this logic on another place, no way to move it to the base component?

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/Address.js:
```diff
@@ -0,0 +1,73 @@
+
+  useEffect(() => {
+    setTimeout(() => setAnchorEl(null), 3000)
+  }, [anchorEl])
```
Why is this necessary?

lcsfelix replied:
This auto closes the 'Copied to clipboard!' popper after 3s. The Material-Ui Popper uses a toggle anchor, so the user would have to click the Copy button again to close the popper. I think it's also a nice touch.

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/tx.js:
```diff
@@ -0,0 +1,12 @@
+    display: 'Bitcoin',
+    code: 'bitcoin',
+    unitScale: 8
+  }
```
Why only btc here?

lcsfelix committed:
fix: change expandable table logic

lcsfelix committed:
fix: add other coins

lcsfelix committed:
refactor: move log download function to component

lcsfelix committed:
refactor: use name values in RadioGroup

RafaelTaranto commented on new-lamassu-admin/src/components/LogsDownloaderPopper.js:
```diff
@@ -121,53 +121,71 @@ const styles = {
-    setRadioButtons(toInteger(event.target.value))
+    const radio = event.target.value
+    setSelectedRadio(radio)
   }
```
this can be reduced to `const handleRadioButtons = compose(setSelectedRadio, get('target.value'))`

RafaelTaranto commented on new-lamassu-admin/src/components/LogsDownloaderPopper.js:
```diff
@@ -121,53 +121,71 @@ const styles = {
-const LogsDownloaderPopover = ({ id, open, anchorEl, onClose, logsResponse, ...props }) => {
-  const [radioButtons, setRadioButtons] = useState(0)
+const LogsDownloaderPopover = ({ id, name, open, anchorEl, getTimestamp, logs, title, ...props }) => {
+  const radioButtonOptions = [{ label: 'All logs', value: 'all' }, { label: 'Date range', value: 'range' }]
```
Maybe create consts for `all` and `range`, that way you can reference that in code instead of `radioButtonOptions[0].value`.

RafaelTaranto commented on new-lamassu-admin/src/components/expandable-table/ExpTable.js:
```diff
@@ -28,26 +27,29 @@ const ExpRow = ({ id, columns, details, sizes, expanded, className, expandRow, .
   const detailsRowClasses = {
     [classes.detailsRow]: true,
-    [classes.hideDetailsRow]: expanded
+    [classes.hideDetailsRow]: !expanded
```
We're not using this object anymore

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/DetailsCard.js:
```diff
@@ -0,0 +1,209 @@
+          </div>
+          <div className={classnames(classes.col, classes.col3)}> {/* Column 3 */}
+            <div className={classnames(classes.innerRow)}>
+              <div style={{ height: 43.4 }}>
```
we should prevent using sytles, why is this div empty?

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/DetailsCard.js:
```diff
@@ -0,0 +1,209 @@
+                      ${addr.slice(35, 40)}
+                      ${addr.slice(40, 45)}
+                      ${addr.slice(45, 50)}`}
+                  </Address>
```
Shouldn't the address component handle the output? Also this won't work for every address, check out the `formatAddress` function on `Funding.js`

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/DetailsCard.js:
```diff
@@ -0,0 +1,209 @@
+                    {txHash &&
+                      `${txHash.slice(0, 41)}
+                      ${txHash.slice(41)}`}
+                  </Address>
```
Why is this using the address component? If it's for both address and txhash shouldn't it be called something else? Something like copytoclipboard or the like.

As far as formatting goes I think it's better to make the div split the hash with break lines instead of doing it by hand.

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/DetailsCard.js:
```diff
@@ -0,0 +1,209 @@
+                  {tx.id &&
+                    `${tx.id.slice(0, 21)}
+                  ${tx.id.slice(21)}`}
+                </Address>
```
Same as above

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/Transactions.js:
```diff
@@ -0,0 +1,206 @@
+
+  const rows = txResponse && txResponse.data.map(tx => {
+    const customerName = tx.customerName ? tx.customerName
+      : (tx.customerIdCardData ? `${startCase(lowerCase(tx.customerIdCardData.firstName.slice(0, 1)))}. ${startCase(lowerCase(tx.customerIdCardData.lastName))}` : tx.customerPhone)
```
This is complicated to read, split into a couple of functions.

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/Transactions.js:
```diff
@@ -0,0 +1,206 @@
+    124, // Time
+    90, // Trade
+    71 // Expand
+  ]
```
I think it makes more sense to pass down the sizes with the main col object

```
 {
          value: moment(tx.created).format('HH:mm:ss'),
          textAlign: 'right',
          size: 12398612
        },
```

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/Transactions.js:
```diff
@@ -0,0 +1,206 @@
+      })
+      FileSaver.saveAs(blob, `${formatDateFile(range.from)}_${formatDateFile(range.to)}_transactions`)
+    }
+  }
```
Didn't we abstracted this away to a component?

lcsfelix committed:
fix: assorted fixes

lcsfelix committed:
feat: virtualize expandable table

lcsfelix committed:
fix: clean up imports

lcsfelix committed:
fix: remove border radius

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/CopyToClipboard.js:
```diff
@@ -0,0 +1,89 @@
+
+  const formatSessionId = (id = '') => {
+    return id.replace(/(.{21})/g, '$1 ')
+  }
```
Still not what I meant. This component should do what it says it does, Copy to Clipboard. Formatting should be handled externally. Format address is something that we need, but the other two are just line breaking at the end of the div size, so you should make it break automatically instead of using a function for this.

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/Transactions.js:
```diff
@@ -0,0 +1,159 @@
+  const formatCustomerName = (customer) => {
+    const { firstName, lastName } = customer
+
+    return `${startCase(lowerCase(firstName.slice(0, 1)))}. ${startCase(lowerCase(lastName))}`
```
Why not do this:

```startCase(lowerCase(`${firstName.slice(0, 1)}. ${lastName}`))```

lcsfelix replied:
`startCase` removes the dot after the first name.

lcsfelix replied:
https://github.com/lamassu/lamassu-server/pull/336/commits/7ecd04765289b4680c8ff903ecee0d0da44af8e6 removes formatting the last name altogether, to prevent misformatting names with special characters, such as [Double-barreled names](https://en.wikipedia.org/wiki/Double-barrelled_name).

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/Transactions.js:
```diff
@@ -0,0 +1,159 @@
+        },
+        {
+          name: 'Address',
+          value: `${tx.toAddress.slice(0, 9)}...`,
```
this can be done with css as well, it's usually better since we don't have to change here if we ever need to change this size.

RafaelTaranto commented on new-lamassu-admin/src/pages/Transactions/Transactions.js:
```diff
@@ -0,0 +1,159 @@
+        },
+        {
+          name: 'Date (UTC)',
+          value: moment(tx.created).format('YYYY-MM-D'),
```
does moment set UTC correctly?

lcsfelix replied:
I assumed the date was already stored in the DB in the proper format, but I can use `moment.utc` to make sure.

lcsfelix committed:
fix: move formatting out of CopyToClipboard

And use CSS instead of JS to format.

lcsfelix commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create transactions page](https://trello.com/c/iEaZLvTE/119-create-transactions-page)

lcsfelix commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create transactions page](https://trello.com/c/iEaZLvTE/119-create-transactions-page)

lcsfelix committed:
fix: remove customer's last name formatting

This was using lodash's string case functions, which produce unwanted
results if, for instance, a user has a double-barrel last name.

RafaelTaranto reviewed (approved):
lgtm

RafaelTaranto commented:
Oh wait, this should be opened to dev.
